src/node_os.cc

```diff
@@ -287,21 +287,29 @@ static void GetUserInfo(const FunctionCallbackInfo<Value>& args) {
     encoding = UTF8;
   }
 
-  const int err = uv_os_get_passwd(&pwd);
-
-  if (err) {
+  if (const int err = uv_os_get_passwd(&pwd)) {
     CHECK_GE(args.Length(), 2);
     env->CollectUVExceptionInfo(args[args.Length() - 1], err,
                                 "uv_os_get_passwd");
     return args.GetReturnValue().SetUndefined();
   }
 
-  auto free_passwd = OnScopeLeave([&]() { uv_os_free_passwd(&pwd); });
+  auto free_passwd = OnScopeLeave([&] { uv_os_free_passwd(&pwd); });
 
   Local<Value> error;
 
+#ifdef _WIN32
+  Local<Value> uid = Number::New(
+      env->isolate(),
+      static_cast<double>(static_cast<int32_t>(pwd.uid & 0xFFFFFFFF)));
```

Comment:
This has ben discussed and I don't want to block this... but this code might be too complicated.

```suggestion
      static_cast<double>(static_cast<int32_t>(pwd.uid)));
```

So that I understand... we start with this...

```C++
typedef struct uv_passwd_s {
    char* username;
    long uid;
    long gid;
    char* shell;
    char* homedir;
} uv_passwd_t;
```

(source: [libuv documentation](https://docs.libuv.org/en/v1.x/misc.html).)


https://github.com/nodejs/node/blob/a243225aa2c016bc02471c5dd3bae3cf906f6d14/deps/uv/include/uv.h#L1212-L1218

Under non-Windows systems.

But, for extra fun... under Windows, we have a different definition...

```
struct uv_passwd_s
{
  char *username;
  unsigned long uid;
  unsigned long gid;
  char *shell;
  char *homedir;
};
```

https://github.com/symplely/uv-ffi/blob/8e225f11537f9b34df7030652d00e1343d457ab1/headers/uv_windows.h#L2997C1-L3004C3


[A long in Visual Studio is effectively an `int32_t`, it goes from -2,147,483,648 to 2,147,483,647](https://learn.microsoft.com/en-us/cpp/cpp/data-type-ranges?view=msvc-170). Whereas an unsigned long is an uint32_t.

Most other compiler systems will have long be either int32_t or int64_t (depending on whether we have a 32-bit or 64-bit system).  They will have long be either uint32_t or uint64_t.

Now, Number::New's function signature takes a double. So the cast to double is mostly just to show what is happening (as it would happen implicitly).


So, setting aside the cast to double, we are doing...

```C++
uint32_t x =...
x = static_cast<int32_t>(x & 0xFFFFFFFF)
```

(If I got something wrong, please tell me.)

What is the purpose of the `0xFFFFFFFF`?

Reply:
I missed these reviews. I'll open another pull-request to apply the changes.

Reply:
@anonrig That's fine. I think your code is correct. It was just nitpicking.

Reply:
The purpose of `& 0xFFFFFFFF` here was to mask higher bits. Unless my knowledge is horribly outdated, casting with overflow here would be UB (even though i doubt existence of implementations that don't just discard extra bits).

No strong opinion on casting to `double`, except for 'explicit > implicit' rule of thumb.

Reply:
> The purpose of & 0xFFFFFFFF here was to mask higher bits. Unless my knowledge is horribly outdated, casting with overflow here would be UB

An unsigned long under Visual Studio is a 32-bit unsigned integer. So we are casting an uint32 to an int32.

https://learn.microsoft.com/en-us/cpp/cpp/data-type-ranges?view=msvc-170

Node is C++20 and in C++20 casting from unsigned to signed and back is well defined.

> No strong opinion on casting to double, except for 'explicit > implicit' rule of thumb.

I do not care at all. I only remarked that @anonrig's code was not perfectly consistent. I did not ask for the code to change.